staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go

```diff
@@ -0,0 +1,225 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package peerproxy
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/client-go/discovery"
+	"k8s.io/klog/v2"
+
+	apidiscoveryv2 "k8s.io/api/apidiscovery/v2"
+	v1 "k8s.io/api/coordination/v1"
+	schema "k8s.io/apimachinery/pkg/runtime/schema"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	responsewriterutil "k8s.io/apiserver/pkg/util/responsewriter"
+)
+
+func (h *peerProxyHandler) addPeerDiscoveryInfo(obj interface{}, identityLeaseLabelSelector string) {
+	serverIdentityLease, ok := isValidIdentityLease(obj, identityLeaseLabelSelector)
+	if !ok {
+		klog.ErrorS(fmt.Errorf("serverIdentityLease invalid"), "error adding peer served resources")
+		return
+	}
+
+	err := h.addPeerDiscoveryInfoToCache(serverIdentityLease)
+	if err != nil {
+		klog.ErrorS(err, "error adding peer served resources", "serverID", serverIdentityLease.Name)
+	}
+}
+
+func (h *peerProxyHandler) deletePeerDiscoveryInfo(obj interface{}, identityLeaseLabelSelector string) {
+	serverIdentityLease, ok := isValidIdentityLease(obj, identityLeaseLabelSelector)
+	if !ok {
+		klog.ErrorS(fmt.Errorf("serverIdentityLease invalid"), "error deleting peer served resources")
+		return
+	}
+
+	h.withPeerAggDiscoveryCacheLock(func(cache map[string]*peerAggDiscoveryInfo) {
+		delete(cache, serverIdentityLease.Name)
+	})
+}
+
+func (h *peerProxyHandler) updatePeerDiscoveryInfo(oldObj, newObj interface{}, identityLeaseLabelSelector string) {
+	oldLease, oldLeaseOk := isValidIdentityLease(oldObj, identityLeaseLabelSelector)
+	newLease, newLeaseOk := isValidIdentityLease(newObj, identityLeaseLabelSelector)
+
+	switch {
+	case !oldLeaseOk && !newLeaseOk:
+		return
+	case !oldLeaseOk && newLeaseOk:
+		if err := h.addPeerDiscoveryInfoToCache(newLease); err != nil {
+			klog.ErrorS(err, "error adding peer to discovery cache", "serverID", newLease.Name)
+		}
+	case oldLeaseOk && !newLeaseOk:
+		h.withPeerAggDiscoveryCacheLock(func(cache map[string]*peerAggDiscoveryInfo) {
+			delete(cache, oldLease.Name)
+		})
+	case oldLeaseOk && newLeaseOk:
+		// Delete old discovery info if holderIdentity changed, implying
+		// the server was restarted.
+		if oldLease.Name == newLease.Name && *oldLease.Spec.HolderIdentity != *newLease.Spec.HolderIdentity {
+			h.withPeerAggDiscoveryCacheLock(func(cache map[string]*peerAggDiscoveryInfo) {
+				delete(cache, oldLease.Name)
+			})
+		}
+		if err := h.addPeerDiscoveryInfoToCache(newLease); err != nil {
+			klog.ErrorS(err, "error adding peer to discovery cache", "serverID", newLease.Name)
+		}
+	}
```

Comment:
this is the event based logic , so you have an event and process it, this implies that you'll have infinite states depending on the chain of events, A->B->C->D ... usually we use a level based logic, so you get an event and you reconcile by checking what is the current state and the desired state, so you can always return to the desired state

Reply:
Resolving since now I switched to using a workqueue that reconstructs peer-discovery-cache from available leases in the cluster everytime a suitable signal is received (suitable signal being - a new peer lease added, a peer lease deleted, a peer lease updated to change the holderIdentity implying a server restart)